File: leetcode/python3/longest-increasing-subsequence.py
#
# @lc app=leetcode id=300 lang=python
#
# [300] Longest Increasing Subsequence
#
class Solution(object):
    def lengthOfLIS(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        # if not nums:
        #     return 0
        # l = [1 for _ in range(len(nums))]
        # for i in range(1, len(nums)):
        #     for j in range(i):
        #         if nums[j] < nums[i]:
        #             l[i] = max(l[i], l[j]+1)
        # return max(l)   
        if not nums:
            return 0
        lst = [nums[0]]
        for n in nums[1:]:
            if n > lst[-1]:
                lst.append(n)
                continue
                
            left, right = 0, len(lst)-1
            while left < right:
                mid = (left + right) >> 1
                if lst[mid] >= n:
                    right = mid
                else:
                    left = mid + 1
            lst[left] = n
        return len(lst)     

File: leetcode/cpp/longest-increasing-subsequence.cpp
/*
 * @lc app=leetcode id=300 lang=cpp
 *
 * [300] Longest Increasing Subsequence
 */

// @lc code=start
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        int n = nums.size();
        if (n < 1) return 0;
        vector<int> tail;
        tail.push_back(nums[0]);
        for (int i = 1; i < n; i++) {
            if (nums[i] > tail.back()) {
                tail.push_back(nums[i]);
                continue;
            }

            int l = 0, r = tail.size()-1;
            while (l < r) {
                int mid = l + r >> 1;
                if (tail[mid] >= nums[i]) r = mid;
                else l = mid + 1;
            }
            tail[l] = nums[i];
        }
        return tail.size();
    }
};
// @lc code=end

